folding-schemes/src/folding/hypernova/mod.rs

```diff
@@ -25,3 +55,534 @@ impl<F: PrimeField> Witness<F> {
+                external_inputs.len(),
+            ));
+        }
+
+        if self.i > C1::ScalarField::from_le_bytes_mod_order(&usize::MAX.to_le_bytes()) {
+            return Err(Error::MaxStep);
+        }
+
+        let mut transcript_p: PoseidonTranscript<C1> =
+            PoseidonTranscript::<C1>::new(&self.poseidon_config.clone());
+        let (nimfs_proof, U_i1, W_i1, rho_bits) = NIMFS::<C1, PoseidonTranscript<C1>>::prove(
+            &mut transcript_p,
+            &self.ccs,
+            &[self.U_i.clone()],
+            &[self.u_i.clone()],
+            &[self.W_i.clone()],
+            &[self.w_i.clone()],
+        )?;
+
+        // sanity check: check the folded instance relation
+        #[cfg(test)]
+        U_i1.check_relation(&self.ccs, &W_i1)?;
+
+        let mut i_bytes: [u8; 8] = [0; 8];
+        i_bytes.copy_from_slice(&self.i.into_bigint().to_bytes_le()[..8]);
+        let i_usize: usize = usize::from_le_bytes(i_bytes);
+
+        let z_i1 = self
+            .F
+            .step_native(i_usize, self.z_i.clone(), external_inputs.clone())?;
+
+        let u_i1_x = U_i1.hash(
+            &self.poseidon_config,
+            self.i + C1::ScalarField::one(),
+            self.z_0.clone(),
+            z_i1.clone(),
+        )?;
+
+        // u_{i+1}.x[1] = H(cf_U_{i+1})
+        let cf_u_i1_x: C1::ScalarField;
+
+        if self.i == C1::ScalarField::zero() {
+            cf_u_i1_x = self.cf_U_i.hash_cyclefold(&self.poseidon_config)?;
+
+            augmented_f_circuit = AugmentedFCircuit::<C1, C2, GC2, FC> {
+                _c2: PhantomData,
+                _gc2: PhantomData,
+                poseidon_config: self.poseidon_config.clone(),
+                ccs: self.ccs.clone(),
+                i: Some(C1::ScalarField::zero()),
+                i_usize: Some(0),
+                z_0: Some(self.z_0.clone()),
+                z_i: Some(self.z_i.clone()),
+                external_inputs: Some(external_inputs.clone()),
+                u_i_C: Some(self.u_i.C),
+                U_i: Some(self.U_i.clone()),
+                U_i1_C: Some(U_i1.C),
+                F: self.F.clone(),
+                x: Some(u_i1_x),
+                nimfs_proof: Some(nimfs_proof),
+
+                // cyclefold values
+                cf_u_i_cmW: None,
+                cf_U_i: None,
+                cf_x: Some(cf_u_i1_x),
+                cf_cmT: None,
+            };
+        } else {
+            let rho_Fq = C2::ScalarField::from_bigint(BigInteger::from_bits_le(&rho_bits))
+                .ok_or(Error::OutOfBounds)?;
+            // CycleFold part:
+            // get the vector used as public inputs 'x' in the CycleFold circuit
+            // cyclefold circuit for cmW
+            let cf_u_i_x = [
+                vec![rho_Fq],
+                get_cm_coordinates(&self.U_i.C),
+                get_cm_coordinates(&self.u_i.C),
+                get_cm_coordinates(&U_i1.C),
+            ]
+            .concat();
```

Comment:
sounds good, although, does it add a lot of computational-complexity/time-to-compute having an `if==F::zero()` there?

Reply:
Well, my main consideration is that against recursive schemes, the only options that folding has are having a blazingly fast folding step such that we can catch up with the recursion overhead at some point.

That means that any branching that we have within this function slows down the overall solution. 
Specially since we need to fold hundreds of thousands of times to say the least.